potato/item_state_management.py

```diff
@@ -0,0 +1,315 @@
+from __future__ import annotations
+
+# Need to import UserState as a type hint for the ItemStateManager
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from user_state_management import UserState
+
+from enum import Enum
+from collections import OrderedDict, deque, Counter, defaultdict
+import random
+
+# Singleton instance of the ItemStateManager
+ITEM_STATE_MANAGER = None
+
+def init_item_state_manager(config: dict) -> ItemStateManager:
+    # TODO: make the manager type configurable between in-memory and DB-backed.
+    # The DB back-end is for when we have a ton of data and don't want it sitting in
+    # memory all the time (or where some external process is going to be adding new items)
+    global ITEM_STATE_MANAGER
+
+    if ITEM_STATE_MANAGER is None:
+        ITEM_STATE_MANAGER = ItemStateManager(config)
+
+    return ITEM_STATE_MANAGER
+
+def get_item_state_manager() -> ItemStateManager:
+    # TODO: make the manager type configurable between in-memory and DB-backed.
+    # The DB back-end is for when we have a ton of data and don't want it sitting in
+    # memory all the time (or where some external process is going to be adding new items)
+    global ITEM_STATE_MANAGER
+
+    if ITEM_STATE_MANAGER is None:
+        raise ValueError("Item State Manager has not been initialized yet!")
+
+    return ITEM_STATE_MANAGER
+
+class Item:
+    """
+    A class for maintaining state on items that are being annotated.
+    The state of the annotations themselves are stored in the UserState class.
+    The item iself is largely immutable but can be updated with metadata.
+    """
+
+    def __init__(self, item_id, item_data):
+        self.item_id = item_id
+        self.item_data = item_data
+        self.metadata = {}
+
+        # This data structure keeps the label-based annotations the user has
+        # completed so far
+        self.labels = {}
+
+        # This data structure keeps the span-based annotations the user has
+        # completed so far
+        self.span_annotations = {}
+
+    def add_metadata(self, metadata_name: str, metadata_value: str):
+        self.metadata[metadata_name] = metadata_value
+
+    def get_id(self):
+        return self.item_id
+
+    def get_data(self):
+        return self.item_data
+
+    def get_metadata(self, metadata_name: str):
+        return self.metadata.get(metadata_name, None)
+
+    def __str__(self):
+        return f"Item(id:{self.item_id}, data:{self.item_data}, metadata:{self.metadata})"
+
+class Label:
```

Comment:
This feels generated and could be simplified to a dataclass given that is stores data for 2 fields. This kinda brings me back to Java OO programming.